### PR TITLE
Proposal for Enabling Multi-Writers Shred Insertion

### PR DESCRIPTION
**Pull request by yhchiang-sol**
## Problem
 The current shred insertion is single-threaded which prevents us from fully
 utilizing the write bandwidth of our RocksDB backed BlockStore.

 The entire shred insertion is protected by an `insert_shreds_lock`, which
 ensures the atomicity of the read-modify-write operations on meta-data
 column families.

## Proposed Solution
This document describes how we can enable multi-writers in shred_insertion
by using RocksDB's merge operator.

### CODE REVIEW

File: docs/src/proposals/multi-writers.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+at the end of the function call to minimize the lock duration.
+
+For the BlockStore, we will utilize RocksDB's merge operator to perform
+read-modify-write operation to avoid data race between different writers.
+
+Specifically, the `insert_shred()` function touches the following column
+families:
+
+### DeadSlots
+DeadSlots stores [(slot, flag)](https://github.com/solana-labs/solana/blob/master/ledger/src/blockstore.rs#L1212)
```
Comment (behzadnouri) [inline docs/src/proposals/multi-writers.md:45]: for links to github, can you please use permalinks instead of links to current master code (which is constantly changing)? for example here the link would be:
https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1212

Comment (yhchiang-sol) [reply docs/src/proposals/multi-writers.md]: Thanks for pointing this out.  Done!

File: docs/src/proposals/multi-writers.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+pairs indicating whether a particular slot can be safely ignored / deleted.
+As it does not involve in any read-modify-write based on its previous state,
+simply using write-batch with put operation is enough to ensure its atomicity
+and thread-safety.
+
+### ShredData / ShredCode
+Both [ShredData](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1473-L1478)
+and [ShredCode](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1274)
+store data directly from the received shred and do not rely on any previous
+metadata.  As a result, they are simply write opreations which do not require
```
Comment (carllin) [inline docs/src/proposals/multi-writers.md:55]: opreations -> operations

File: docs/src/proposals/multi-writers.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+merge operation to ensure potential read-modify-write atomicity.
+
+### SlotMeta
+The SlotMeta column family involves in intensive read-modify-write operation
+and requires converting to merge operator in order to ensure atomicity without
+locking.  Here're how we can ensure thread-safety for each field:
```
Comment (carllin) [inline docs/src/proposals/multi-writers.md:61]: A bit concerned with races with the write to `SlotMeta` here in  `clear_unconfirmed_slot()`: https://github.com/solana-labs/solana/blob/45e56c599d726766f7c4440a73ff29b77fa5abf6/ledger/src/blockstore.rs#L987-L989. If this function writes `SlotMeta` version `A` while a merge is also happening which writes version `B`, is it possible that the merge operation version `B` wins here, *after* `clear_unconfirmed_slot()` has already cleared the rest of the column families? 

This would be bad because then we would possibly have the merged `SlotMeta`, but the data + coding shreds + other metadata have been deleted by `clear_unconfirmed_slot()`, which may prevent things like insertion of future data + coding shreds

Comment (carllin) [reply docs/src/proposals/multi-writers.md:61]: If these races are hard to work around, we can keep an `RwLock` around to make sure the `insert` path and `clear_unconfirmed_slot()` remain mutually exclusive in v1.

File: docs/src/proposals/multi-writers.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+As it does not involve in any read-modify-write based on its previous state,
+simply using write-batch with put operation is enough to ensure its atomicity
+and thread-safety.
+
+### ShredData / ShredCode
+Both [ShredData](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1473-L1478)
+and [ShredCode](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1274)
+store data directly from the received shred and do not rely on any previous
+metadata.  As a result, they are simply write opreations which do not require
+merge operation to ensure potential read-modify-write atomicity.
```
Comment (carllin) [inline docs/src/proposals/multi-writers.md:56]: Only concerns I would have here:

1) Races with `clear_unconfirmed_slot()`: https://github.com/solana-labs/solana/blob/45e56c599d726766f7c4440a73ff29b77fa5abf6/ledger/src/blockstore.rs#L980-L982, which is removing the slot data and coding shreds. If a deletion is issued, this will not cause any corruption right? If a deletion happens during the middle of a merge, worst case the merge is just writing stale data (this is probably acceptable)?

2) Will the merge updates to the column families happen in an atomic transaction? Temporary inconsistency between column families might break some inconsistencies. For instance if the SlotMeta merge happens first, we may have a "consumed" value of `N`, but the shred for `N` may not have been written to blockstore yet. Or perhaps the index hasn't been updated yet for that slot, which would cause code like this to panic: https://github.com/solana-labs/solana/blob/45e56c599d726766f7c4440a73ff29b77fa5abf6/core/src/repair_generic_traversal.rs#L59-L61

3. Also, the validator may crash during the merge operation for one of these column families. For instance if we do the merge operation on the SlotMeta first, and then we crash during the merge of the ShredData, these two column families will be out of sync on restart. Will we need some consistency checkers?

File: docs/src/proposals/multi-writers.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+#### [last_index](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L26-L27)
+The last_index field doesn't require merge operator to ensure its atomicity as
+there should be only one shred which has its [is_last_in_slot to be true]
+(https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3189-L3200).
+
+#### [parent_slot](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L28-L29)
+parent_slot is updated only from [std::u64::MAX to a valid parent slot](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3251-L3253).
+A merge operator is needed to ensure we never override a valid parent slot
+by u64::MAX.
```
Comment (carllin) [inline docs/src/proposals/multi-writers.md:92]: Just an fyi: Note it's possible that a malicious leader could send two shreds for the same slot with a different `parent_slot` field. We don't currently handle that in the checks either 😃 https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1162. 

Right now we take the parent of the first shred we receive for the slot `N`. So if the first shred in slot `N` has the incorrect parent, there will be an error in replay (because the PoH hash verification chain will fail from the wrong parent), and the validator will mark the slot `N` as dead. Other validators that get the correct parent first will successfully replay.

The only way this partition is successfully resolved is if each validator sees the alternate version of the duplicate shred and calls `handle_duplicate` here: https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1197

File: docs/src/proposals/multi-writers.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+inside [`chain_new_slot_to_prev_slot`](https://github.com/yhchiang-sol/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3543-L3550).
+As we never remove entries from the next_slots field, its merge operator can be
+implemented by taking the union of the new value and the existing value.
+
+#### [is_connected](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L33-L35)
+`is_connected` field is either updated based on
+[the previous slot](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3535)
+or via [propagation](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3469).
+As `is_connected` is never set from true to false, its merge operator can be
+implemented using the OR operation.
```
Comment (carllin) [inline docs/src/proposals/multi-writers.md:105]: heh this field is already so busted and unused I'm temped to remove it and all the complicated associated logic. I think it's probably broken in the majority of validators that started from a snapshot. 

The reason being when you start from a snapshot of slot `X`, you unpack all the account state for slot `X`, but `X` does not actually exist in the ledger (You don't need the transactions from `X` since the snapshot already contains all of the state). 

Then all blockstore slots will descend from `X` because it's the root block, so blockstore insert will insert an empty `SlotMeta` for `X` whenever it inserts a shred from a child slot of `X` (kind of like what we do now if you get a shrd from the child slot before the parent slot). This empty `SlotMeta` for `X` will have `is_connected == false`, and so all children will have `is_connected == false` as well

File: docs/src/proposals/multi-writers.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+#### [completed_data_indexes](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L36-L37)
+The `completed_data_indexes` remember the indexes which belonging slots
+are completed inside and are updated inside
+[`update_completed_data_indexes`](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3132-L3168).
+Its merge operator can simply burrow `update_completed_data_indexes`.
```
Comment (carllin) [inline docs/src/proposals/multi-writers.md:111]: burrow -> borrow

Comment (carllin) [reply docs/src/proposals/multi-writers.md:111]: Will the merge operator will have access to data from other column families as well? Needs the `ShredIndex` to figure out if all shreds in a range have been completed: https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3164. Can we guarantee that this merge operation happens *after* the merge operation on the `ShredIndex` is done? Ordering seems important here

**Comment by yhchiang-sol** [comment]
Thanks for the detailed comments, @carllin.  I will take a deeper look into your comments and the interaction between `insert_shreds` and `clear_unconfirmed_slot`.  Here're some information and clarification about the merge operator which might answer some of your questions/concerns.

You can think of the merge operator as a smarter write who can also do something at the read time.

For instance, consider the following example where the value that we persist is a vector, and we write a merge operation that takes the union.  Then if we have the following pattern:

```
db.put(key, {"a", "b"});
db.merge(key, "c");
```

What stores in RocksDB will look something like the following:

```
merge operand: "c"
value entry:  {"a", "b"}
```

Then when do `db.get(key)`, RocksDB will run the merge operator on the fly, which will essentially return `{"a", "b", "c"}` to  us.

Consider another example where we store a boolean.  Let's say:
```
db.put(key, true);
db.put(key, false);
```
In this case, the end value will be `false`.  But if we implement a merge operator where we perform OR operation, then consider the following:
```
db.merge(key, true);
db.merge(key, false);
```
Then when we call `db.get(key)`, we will get `true` value instead.  In other words, it can allow a certain level of racing but resolve it later at the read time (and compaction as well) with our custom merge operator logic.
In other words, the following two code is equivalent:
```
mutex lock;
bool old_value = db.get(key);
value.put(key, value | new_value);
```
And
```
value.merge(key, new_value);   // and we implement the merge operator by taking the OR operation.
```

**Comment by carllin** [comment]
@yhchiang-sol thanks, our chat was clarifying! 

The most important takeaway was that if multiple column families' merge inserts are put into a write batch, the atomicity of the merge updates to all the column families at the same time is maintained by any readers

**Comment by behzadnouri** [comment]
One question is that if something can be done (e.g. a wrapper with specific api, or anything else), that makes this future-proof?

I mean lets say all current uses are vetted or fixed so that there is no race condition. Still, that will not guarantee that a future code change unknowingly adds a subtle race condition; and in particular no practical amount of testing will exhaustively cover all scenarios causing race conditions.

**Comment by yhchiang-sol** [comment]
> One question is that if something can be done (e.g. a wrapper with specific api, or anything else), that makes this future-proof?
> I mean lets say all current uses are vetted or fixed so that there is no race condition. Still, that will not guarantee that a future code change unknowingly adds a subtle race condition; and in particular no practical amount of testing will exhaustively cover all scenarios causing race conditions.

Good question.  At the minimum, A set of new tests for each column family for each field will be added to ensure the correctness under different insertion order. 

As for a general way to detect whether unwanted racing happens, we might be able to detect them by adding checks inside the merge operator logic and report inconsistency during the read time via error message or counter, and we need to define what is our consistency guarantee.
